<center>
    <img src="https://gitlab.com/ibm/skills-network/courses/placeholder101/-/raw/master/labs/module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo"  />
</center>


# **SpaceX  Falcon 9 first stage Landing Prediction**


# Lab 1: Collecting the data


Estimated time needed: **45** minutes


In this capstone, we will predict if the Falcon 9 first stage will land successfully. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because SpaceX can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against SpaceX for a rocket launch. In this lab, you will collect and make sure the data is in the correct format from an API. The following is an example of a successful and launch.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/landing\_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/crash.gif)


Most unsuccessful landings are planned. Space X performs a controlled landing in the oceans.


## Objectives


In this lab, you will make a get request to the SpaceX API. You will also do some basic data wrangling and formating.

*   Request to the SpaceX API
*   Clean the requested data


***


## Import Libraries and Define Auxiliary Functions


We will import the following libraries into the lab


In [2]:
# Requests allows us to make HTTP requests which we will use to get data from an API
import requests

In [3]:
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd

In [4]:
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np

In [5]:
# Datetime is a library that allows us to represent dates
import datetime

In [6]:
# Setting this option will print all collumns of a dataframe
pd.set_option('display.max_columns', None)
# Setting this option will print all of the data in a feature
pd.set_option('display.max_colwidth', None)

Below we will define a series of helper functions that will help us use the API to extract information using identification numbers in the launch data.

From the <code>rocket</code> column we would like to learn the booster name.


In [7]:
# Takes the dataset and uses the rocket column to call the API and append the data to the list
def getBoosterVersion(data):
    for x in data['rocket']:
        response = requests.get("https://api.spacexdata.com/v4/rockets/"+str(x)).json()
        BoosterVersion.append(response['name'])

From the <code>launchpad</code> we would like to know the name of the launch site being used, the logitude, and the latitude.


In [8]:
# Takes the dataset and uses the launchpad column to call the API and append the data to the list
def getLaunchSite(data):
    for x in data['launchpad']:
        response = requests.get("https://api.spacexdata.com/v4/launchpads/"+str(x)).json()
        Longitude.append(response['longitude'])
        Latitude.append(response['latitude'])
        LaunchSite.append(response['name'])

From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to.


In [9]:
# Takes the dataset and uses the payloads column to call the API and append the data to the lists
def getPayloadData(data):
    for load in data['payloads']:
        response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
        PayloadMass.append(response['mass_kg'])
        Orbit.append(response['orbit'])

From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, wheter the core is reused, wheter legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.


In [10]:
# Takes the dataset and uses the cores column to call the API and append the data to the lists
def getCoreData(data):
    for core in data['cores']:
            if core['core'] != None:
                response = requests.get("https://api.spacexdata.com/v4/cores/"+core['core']).json()
                Block.append(response['block'])
                ReusedCount.append(response['reuse_count'])
                Serial.append(response['serial'])
            else:
                Block.append(None)
                ReusedCount.append(None)
                Serial.append(None)
            Outcome.append(str(core['landing_success'])+' '+str(core['landing_type']))
            Flights.append(core['flight'])
            GridFins.append(core['gridfins'])
            Reused.append(core['reused'])
            Legs.append(core['legs'])
            LandingPad.append(core['landpad'])

Now let's start requesting rocket launch data from SpaceX API with the following URL:


In [11]:
spacex_url="https://api.spacexdata.com/v4/launches/past"

In [12]:
response = requests.get(spacex_url)

Check the content of the response


In [14]:
response.status_code

200

You should see the response contains massive information about SpaceX launches. Next, let's try to discover some more relevant information for this project.


### Task 1: Request and parse the SpaceX launch data using the GET request


To make the requested JSON results more consistent, we will use the following static response object for this project:


In [15]:
static_json_url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'

We should see that the request was successfull with the 200 status response code


In [16]:
response.status_code

200

Now we decode the response content as a Json using <code>.json()</code> and turn it into a Pandas dataframe using <code>.json_normalize()</code>


In [17]:
# Use json_normalize meethod to convert the json result into a dataframe
data = pd.json_normalize(response.json())
data.head()

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   
1                      None                    NaN  False     0.0   
2                      None                    NaN  False     0.0   
3  2008-09-20T00:00:00.000Z           1.221869e+09  False     0.0   
4                      None                    NaN  False     0.0   

                     rocket  success  \
0  5e9d0d95eda69955f709d1eb    False   
1  5e9d0d95eda69955f709d1eb    False   
2  5e9d0d95eda69955f709d1eb    False   
3  5e9d0d95eda69955f709d1eb     True   
4  5e9d0d95eda69955f709d1eb     True   

                                                                                                            failures  \
0                                                [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   
1            [{'time': 301, 'altitude': 289, 'reason': 'harmonic oscillation leading to premature engine shutdown'}]   
2  [{'time': 140, 'altitude': 35, 'reason': 'residual stage-1 thrust led to collision between stage 1 and stage 2'}]   
3                                                                                                                 []   
4                                                                                                                 []   

                                                                                                                                                                                details  \
0                                                                                                                                      Engine failure at 33 seconds and loss of vehicle   
1  Successful first stage burn and transition to second stage, maximum altitude 289 km, Premature engine shutdown at T+7 min 30 s, Failed to reach orbit, Failed to recover first stage   
2                                                                                                                  Residual stage 1 thrust led to collision between stage 1 and stage 2   
3                        Ratsat was carried to orbit on the first successful orbital launch of any privately funded and developed, liquid-propelled carrier rocket, the SpaceX Falcon 1   
4                                                                                                                                                                                  None   

  crew ships capsules                                              payloads  \
0   []    []       []                            [5eb0e4b5b6c3bb0006eeb1e1]   
1   []    []       []                            [5eb0e4b6b6c3bb0006eeb1e2]   
2   []    []       []  [5eb0e4b6b6c3bb0006eeb1e3, 5eb0e4b6b6c3bb0006eeb1e4]   
3   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e5]   
4   []    []       []                            [5eb0e4b7b6c3bb0006eeb1e6]   

                  launchpad  flight_number         name  \
0  5e9e4502f5090995de566f86              1    FalconSat   
1  5e9e4502f5090995de566f86              2      DemoSat   
2  5e9e4502f5090995de566f86              3  Trailblazer   
3  5e9e4502f5090995de566f86              4       RatSat   
4  5e9e4502f5090995de566f86              5     RazakSat   

                   date_utc   date_unix                 date_local  \
0  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   
1  2007-03-21T01:10:00.000Z  1174439400  2007-03-21T13:10:00+12:00   
2  2008-08-03T03:34:00.000Z  1217734440  2008-08-03T15:34:00+12:00   
3  2008-09-28T23:15:00.000Z  1222643700  2008-09-28T11:15:00+12:00   
4  2009-07-13T03:35:00.000Z  1247456100  2009-07-13T15:35:00+12:00   

  date_precision  upcoming  \
0           hour     False   
1           hour     False   
2           hour     False   
3           hour     False   
4           hour     False   

                                                                           

Using the dataframe <code>data</code> print the first 5 rows


In [18]:
# Get the head of the dataframe
data.head(1)

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   

                     rocket  success  \
0  5e9d0d95eda69955f709d1eb    False   

                                                              failures  \
0  [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   

                                            details crew ships capsules  \
0  Engine failure at 33 seconds and loss of vehicle   []    []       []   

                     payloads                 launchpad  flight_number  \
0  [5eb0e4b5b6c3bb0006eeb1e1]  5e9e4502f5090995de566f86              1   

        name                  date_utc   date_unix                 date_local  \
0  FalconSat  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   

  date_precision  upcoming  \
0           hour     False   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   auto_update    tbd launch_library_id                        id  \
0         True  False              None  5eb87cd9ffd86e000604b32a   

  fairings.reused fairings.recovery_attempt fairings.recovered fairings.ships  \
0           False                     False              False             []   

                                 links.patch.small  \
0  https://images2.imgbox.com/3c/0e/T8iJcSN3_o.png   

                                 links.patch.large links.reddit.campaign  \
0  https://images2.imgbox.com/40/e3/GypSkayF_o.png                  None   

  links.reddit.launch links.reddit.media links.reddit.recovery  \
0                None               None                  None   

  links.flickr.small links.flickr.original links.presskit  \
0                 []                    []           None   

                                 links.webcast links.youtube_id  \
0  https://www.youtube.com/watch?v=0a_00nJ_Y88      0a_00nJ_Y88   

                                                                  links.article  \
0  https://www.space.com/2196-spacex-inaugural-falcon-1-rocket-lost-launch.html   

                         links.wikipedia  fairings  
0  https://en.wikipedia.org/wiki/DemoSat       NaN

In [19]:
data.columns

Index(['static_fire_date_utc', 'static_fire_date_unix', 'net', 'window',
       'rocket', 'success', 'failures', 'details', 'crew', 'ships', 'capsules',
       'payloads', 'launchpad', 'flight_number', 'name', 'date_utc',
       'date_unix', 'date_local', 'date_precision', 'upcoming', 'cores',
       'auto_update', 'tbd', 'launch_library_id', 'id', 'fairings.reused',
       'fairings.recovery_attempt', 'fairings.recovered', 'fairings.ships',
       'links.patch.small', 'links.patch.large', 'links.reddit.campaign',
       'links.reddit.launch', 'links.reddit.media', 'links.reddit.recovery',
       'links.flickr.small', 'links.flickr.original', 'links.presskit',
       'links.webcast', 'links.youtube_id', 'links.article', 'links.wikipedia',
       'fairings'],
      dtype='object')

You will notice that a lot of the data are IDs. For example the rocket column has no information about the rocket just an identification number.

We will now use the API again to get information about the launches using the IDs given for each launch. Specifically we will be using columns <code>rocket</code>, <code>payloads</code>, <code>launchpad</code>, and <code>cores</code>.


In [20]:
df = data[data['cores'].map(len)==3][['cores']]
df['length'] = df['cores'].map(lambda x : len(x))
df['length2'] = df['cores'].map(len)
df

cores  \
54   [{'core': '5e9e28a5f359187f703b2653', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': False, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}, {'core': '5e9e28a2f359187f273b2642', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb90a834e7c8'}, {'core': '5e9e28a2f3591845c73b2640', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb267a34e7c7'}]   
76  [{'core': '5e9e28a6f3591897453b265f', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}, {'core': '5e9e28a6f359183c413b265d', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb267a34e7c7'}, {'core': '5e9e28a6f359188fd53b265e', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb90a834e7c8'}]   
80   [{'core': '5e9e28a7f3591878063b2661', 'flight': 1, 'gridfins': True, 'legs': True, 'reused': False, 'landing_attempt': True, 'landing_success': False, 'landing_type': 'ASDS', 'landpad': '5e9e3032383ecb6bb234e7ca'}, {'core': '5e9e28a6f359183c413b265d', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb267a34e7c7'}, {'core': '5e9e28a6f359188fd53b265e', 'flight': 2, 'gridfins': True, 'legs': True, 'reused': True, 'landing_attempt': True, 'landing_success': True, 'landing_type': 'RTLS', 'landpad': '5e9e3032383ecb90a834e7c8'}]   

    length  length2  
54       3        3  
76       3        3  
80       3        3

In [21]:
data['cores']

0                           [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]
1                           [{'core': '5e9e289ef35918416a3b2624', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]
2                           [{'core': '5e9e289ef3591814873b2625', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]
3                           [{'core': '5e9e289ef3591855dc3b2626', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]
4                           [{'core': '5e9e289ef359184f103b2627', 'flight': 1, 'gridfins': False, 'legs': False, 're

In [22]:
# Lets take a subset of our dataframe keeping only the features we want and the flight number, and date_utc.
data = data[['rocket', 'payloads', 'launchpad', 'cores', 'flight_number', 'date_utc']]

# We will remove rows with multiple cores because those are falcon rockets with 2 extra rocket boosters and rows that have multiple payloads in a single rocket.
data = data[data['cores'].map(len)==1]
data = data[data['payloads'].map(len)==1]

# Since payloads and cores are lists of size 1 we will also extract the single value in the list and replace the feature.
data['cores'] = data['cores'].map(lambda x : x[0])
data['payloads'] = data['payloads'].map(lambda x : x[0])

# We also want to convert the date_utc to a datetime datatype and then extracting the date leaving the time
data['date'] = pd.to_datetime(data['date_utc']).dt.date

# Using the date we will restrict the dates of the launches
data = data[data['date'] <= datetime.date(2020, 11, 13)]

*   From the <code>rocket</code> we would like to learn the booster name

*   From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to

*   From the <code>launchpad</code> we would like to know the name of the launch site being used, the longitude, and the latitude.

*   From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, whether the core is reused, whether legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.

The data from these requests will be stored in lists and will be used to create a new dataframe.


In [23]:
#Global variables 
BoosterVersion = []
PayloadMass = []
Orbit = []
LaunchSite = []
Outcome = []
Flights = []
GridFins = []
Reused = []
Legs = []
LandingPad = []
Block = []
ReusedCount = []
Serial = []
Longitude = []
Latitude = []

These functions will apply the outputs globally to the above variables. Let's take a looks at <code>BoosterVersion</code> variable. Before we apply  <code>getBoosterVersion</code> the list is empty:


In [24]:
BoosterVersion

[]

Now, let's apply <code> getBoosterVersion</code> function method to get the booster version


In [25]:
# Call getBoosterVersion
getBoosterVersion(data)

the list has now been update


In [26]:
BoosterVersion[0:5]

['Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 9']

we can apply the rest of the  functions here:


In [27]:
# Call getLaunchSite
getLaunchSite(data)

In [28]:
# Call getPayloadData
getPayloadData(data)

In [29]:
# Call getCoreData
getCoreData(data)

Finally lets construct our dataset using the data we have obtained. We we combine the columns into a dictionary.


In [30]:
launch_dict = {'FlightNumber': list(data['flight_number']),
'Date': list(data['date']),
'BoosterVersion':BoosterVersion,
'PayloadMass':PayloadMass,
'Orbit':Orbit,
'LaunchSite':LaunchSite,
'Outcome':Outcome,
'Flights':Flights,
'GridFins':GridFins,
'Reused':Reused,
'Legs':Legs,
'LandingPad':LandingPad,
'Block':Block,
'ReusedCount':ReusedCount,
'Serial':Serial,
'Longitude': Longitude,
'Latitude': Latitude}


Then, we need to create a Pandas data frame from the dictionary launch_dict.


In [31]:
launch_dict.keys()

dict_keys(['FlightNumber', 'Date', 'BoosterVersion', 'PayloadMass', 'Orbit', 'LaunchSite', 'Outcome', 'Flights', 'GridFins', 'Reused', 'Legs', 'LandingPad', 'Block', 'ReusedCount', 'Serial', 'Longitude', 'Latitude'])

In [39]:
import json

In [44]:
print(json.dumps(launch_dict, default = str, indent = 4))

{
    "FlightNumber": [
        1,
        2,
        4,
        5,
        6,
        8,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50,
        51,
        52,
        53,
        54,
        57,
        58,
        59,
        60,
        61,
        63,
        64,
        65,
        66,
        67,
        68,
        69,
        70,
        71,
        72,
        73,
        74,
        76,
        78,
        79,
        80,
        82,
        83,
        84,
        85,
        86,
        87,
        88,
        89,
        90,
        91,
        92,
        93

In [37]:
for key, value in launch_dict.items():
    print(key , ' - ',len(value),' - ', value,'\n')

FlightNumber  -  94  -  [1, 2, 4, 5, 6, 8, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 22, 23, 24, 25, 26, 27, 28, 29, 30, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 57, 58, 59, 60, 61, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 76, 78, 79, 80, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 100, 101, 102, 103, 104, 105, 106] 

Date  -  94  -  [datetime.date(2006, 3, 24), datetime.date(2007, 3, 21), datetime.date(2008, 9, 28), datetime.date(2009, 7, 13), datetime.date(2010, 6, 4), datetime.date(2012, 5, 22), datetime.date(2013, 3, 1), datetime.date(2013, 9, 29), datetime.date(2013, 12, 3), datetime.date(2014, 1, 6), datetime.date(2014, 4, 18), datetime.date(2014, 7, 14), datetime.date(2014, 8, 5), datetime.date(2014, 9, 7), datetime.date(2014, 9, 21), datetime.date(2015, 1, 10), datetime.date(2015, 2, 11), datetime.date(2015, 4, 14), datetime.date(2015, 4, 27), datetime.date(2015, 6, 28), datetime.date(2015, 12, 

In [77]:
# Create a data from launch_dict
pd.set_option("display.max_rows", None)
pd.set_option('display.max_columns',None)
launch_df = pd.DataFrame(launch_dict)

Show the summary of the dataframe


In [79]:
# Show the head of the dataframe
launch_df

FlightNumber        Date BoosterVersion  PayloadMass  Orbit  \
0              1  2006-03-24       Falcon 1        20.00    LEO   
1              2  2007-03-21       Falcon 1          NaN    LEO   
2              4  2008-09-28       Falcon 1       165.00    LEO   
3              5  2009-07-13       Falcon 1       200.00    LEO   
4              6  2010-06-04       Falcon 9          NaN    LEO   
5              8  2012-05-22       Falcon 9       525.00    LEO   
6             10  2013-03-01       Falcon 9       677.00    ISS   
7             11  2013-09-29       Falcon 9       500.00     PO   
8             12  2013-12-03       Falcon 9      3170.00    GTO   
9             13  2014-01-06       Falcon 9      3325.00    GTO   
10            14  2014-04-18       Falcon 9      2296.00    ISS   
11            15  2014-07-14       Falcon 9      1316.00    LEO   
12            16  2014-08-05       Falcon 9      4535.00    GTO   
13            17  2014-09-07       Falcon 9      4428.00    GTO   
14            18  2014-09-21       Falcon 9      2216.00    ISS   
15            19  2015-01-10       Falcon 9      2395.00    ISS   
16            20  2015-02-11       Falcon 9       570.00  ES-L1   
17            22  2015-04-14       Falcon 9      1898.00    ISS   
18            23  2015-04-27       Falcon 9      4707.00    GTO   
19            24  2015-06-28       Falcon 9      2477.00    ISS   
20            25  2015-12-22       Falcon 9      2034.00    LEO   
21            26  2016-01-17       Falcon 9       553.00     PO   
22            27  2016-03-04       Falcon 9      5271.00    GTO   
23            28  2016-04-08       Falcon 9      3136.00    ISS   
24            29  2016-05-06       Falcon 9      4696.00    GTO   
25            30  2016-05-27       Falcon 9      3100.00    GTO   
26            32  2016-07-18       Falcon 9      2257.00    ISS   
27            33  2016-08-14       Falcon 9      4600.00    GTO   
28            34  2016-09-01       Falcon 9      5500.00    GTO   
29            35  2017-01-14       Falcon 9      9600.00     PO   
30            36  2017-02-19       Falcon 9      2490.00    ISS   
31            37  2017-03-16       Falcon 9      5600.00    GTO   
32            38  2017-03-30       Falcon 9      5300.00    GTO   
33            39  2017-05-01       Falcon 9          NaN    LEO   
34            40  2017-05-15       Falcon 9      6070.00    GTO   
35            41  2017-06-03       Falcon 9      2708.00    ISS   
36            42  2017-06-23       Falcon 9      3669.00    GTO   
37            43  2017-06-25       Falcon 9      9600.00     PO   
38            44  2017-07-05       Falcon 9      6761.00    GTO   
39            45  2017-08-14       Falcon 9      2910.00    ISS   
40            46  2017-08-24       Falcon 9       475.00    SSO   
41            47  2017-09-07       Falcon 9      4990.00    LEO   
42            48  2017-10-09       Falcon 9      9600.00     PO   
43            49  2017-10-11       Falcon 9      5200.00    GTO   
44            50  2017-10-30       Falcon 9      3700.00    GTO   
45            51  2017-12-15       Falcon 9      2205.00    ISS   
46            52  2017-12-23       Falcon 9      9600.00     PO   
47            53  2018-01-08       Falcon 9          NaN    LEO   
48            54  2018-01-31       Falcon 9      4230.00    GTO   
49            57  2018-03-06       Falcon 9      6092.00    GTO   
50            58  2018-03-30       Falcon 9      9600.00     PO   
51            59  2018-04-02       Falcon 9      2760.00    ISS   
52            60  2018-04-18       Falcon 9       350.00    HEO   
53            61  2018-05-11       Falcon 9      3750.00    GTO   
54            63  2018-06-04       Falcon 9      5383.85    GTO   
55            64  2018-06-29       Falcon 9      2410.00    ISS   
56            65  2018-07-22       Falcon 9      7076.00    GTO   
57            66  2018-07-25       Falcon 9      9600.00     PO   
58            67  2018-08-07       Falcon 9      58

### Task 2: Filter the dataframe to only include `Falcon 9` launches


Finally we will remove the Falcon 1 launches keeping only the Falcon 9 launches. Filter the data dataframe using the <code>BoosterVersion</code> column to only keep the Falcon 9 launches. Save the filtered data to a new dataframe called <code>data_falcon9</code>.


In [78]:
# Hint data['BoosterVersion']!='Falcon 1'
print(launch_df['BoosterVersion'].value_counts())
data_falcon9 = launch_df[launch_df['BoosterVersion'] != 'Falcon 1']
data_falcon9

Falcon 9    90
Falcon 1     4
Name: BoosterVersion, dtype: int64


FlightNumber        Date BoosterVersion  PayloadMass  Orbit    LaunchSite  \
4              6  2010-06-04       Falcon 9          NaN    LEO  CCSFS SLC 40   
5              8  2012-05-22       Falcon 9       525.00    LEO  CCSFS SLC 40   
6             10  2013-03-01       Falcon 9       677.00    ISS  CCSFS SLC 40   
7             11  2013-09-29       Falcon 9       500.00     PO   VAFB SLC 4E   
8             12  2013-12-03       Falcon 9      3170.00    GTO  CCSFS SLC 40   
9             13  2014-01-06       Falcon 9      3325.00    GTO  CCSFS SLC 40   
10            14  2014-04-18       Falcon 9      2296.00    ISS  CCSFS SLC 40   
11            15  2014-07-14       Falcon 9      1316.00    LEO  CCSFS SLC 40   
12            16  2014-08-05       Falcon 9      4535.00    GTO  CCSFS SLC 40   
13            17  2014-09-07       Falcon 9      4428.00    GTO  CCSFS SLC 40   
14            18  2014-09-21       Falcon 9      2216.00    ISS  CCSFS SLC 40   
15            19  2015-01-10       Falcon 9      2395.00    ISS  CCSFS SLC 40   
16            20  2015-02-11       Falcon 9       570.00  ES-L1  CCSFS SLC 40   
17            22  2015-04-14       Falcon 9      1898.00    ISS  CCSFS SLC 40   
18            23  2015-04-27       Falcon 9      4707.00    GTO  CCSFS SLC 40   
19            24  2015-06-28       Falcon 9      2477.00    ISS  CCSFS SLC 40   
20            25  2015-12-22       Falcon 9      2034.00    LEO  CCSFS SLC 40   
21            26  2016-01-17       Falcon 9       553.00     PO   VAFB SLC 4E   
22            27  2016-03-04       Falcon 9      5271.00    GTO  CCSFS SLC 40   
23            28  2016-04-08       Falcon 9      3136.00    ISS  CCSFS SLC 40   
24            29  2016-05-06       Falcon 9      4696.00    GTO  CCSFS SLC 40   
25            30  2016-05-27       Falcon 9      3100.00    GTO  CCSFS SLC 40   
26            32  2016-07-18       Falcon 9      2257.00    ISS  CCSFS SLC 40   
27            33  2016-08-14       Falcon 9      4600.00    GTO  CCSFS SLC 40   
28            34  2016-09-01       Falcon 9      5500.00    GTO  CCSFS SLC 40   
29            35  2017-01-14       Falcon 9      9600.00     PO   VAFB SLC 4E   
30            36  2017-02-19       Falcon 9      2490.00    ISS    KSC LC 39A   
31            37  2017-03-16       Falcon 9      5600.00    GTO    KSC LC 39A   
32            38  2017-03-30       Falcon 9      5300.00    GTO    KSC LC 39A   
33            39  2017-05-01       Falcon 9          NaN    LEO    KSC LC 39A   
34            40  2017-05-15       Falcon 9      6070.00    GTO    KSC LC 39A   
35            41  2017-06-03       Falcon 9      2708.00    ISS    KSC LC 39A   
36            42  2017-06-23       Falcon 9      3669.00    GTO    KSC LC 39A   
37            43  2017-06-25       Falcon 9      9600.00     PO   VAFB SLC 4E   
38            44  2017-07-05       Falcon 9      6761.00    GTO    KSC LC 39A   
39            45  2017-08-14       Falcon 9      2910.00    ISS    KSC LC 39A   
40            46  2017-08-24       Falcon 9       475.00    SSO   VAFB SLC 4E   
41            47  2017-09-07       Falcon 9      4990.00    LEO    KSC LC 39A   
42            48  2017-10-09       Falcon 9      9600.00     PO   VAFB SLC 4E   
43            49  2017-10-11       Falcon 9      5200.00    GTO    KSC LC 39A   
44            50  2017-10-30       Falcon 9      3700.00    GTO    KSC LC 39A   
45            51  2017-12-15       Falcon 9      2205.00    ISS  CCSFS SLC 40   
46            52  2017-12-23       Falcon 9      9600.00     PO   VAFB SLC 4E   
47            53  2018-01-08       Falcon 9          NaN    LEO  CCSFS SLC 40   
48            54  2018-01-31       Falcon 9      4230.00    GTO  CCSFS SLC 40   
49            57  2018-03-06       Falcon 9      6092.00    GTO  CCSFS SLC 40   
50            58  2018-03-30       Falcon 9      9600.00     PO   VAFB SLC 4E   
51            59  2018-04-02       Falcon 9      2760.00    ISS  CCSFS SLC 40   
52            60  2018-04-18       

Now that we have removed some values we should reset the FlgihtNumber column


In [81]:
(data_falcon9['FlightNumber'] == data_falcon9.loc[:,'FlightNumber']).value_counts()

True    90
Name: FlightNumber, dtype: int64

In [90]:
data_falcon9['FlightNumber'] = list(range(1,data_falcon9['FlightNumber'].shape[0]+1))
data_falcon9

/opt/conda/envs/Python-3.7-OpenCE/lib/python3.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


FlightNumber        Date BoosterVersion  PayloadMass  Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9          NaN    LEO  CCSFS SLC 40   
5              2  2012-05-22       Falcon 9       525.00    LEO  CCSFS SLC 40   
6              3  2013-03-01       Falcon 9       677.00    ISS  CCSFS SLC 40   
7              4  2013-09-29       Falcon 9       500.00     PO   VAFB SLC 4E   
8              5  2013-12-03       Falcon 9      3170.00    GTO  CCSFS SLC 40   
9              6  2014-01-06       Falcon 9      3325.00    GTO  CCSFS SLC 40   
10             7  2014-04-18       Falcon 9      2296.00    ISS  CCSFS SLC 40   
11             8  2014-07-14       Falcon 9      1316.00    LEO  CCSFS SLC 40   
12             9  2014-08-05       Falcon 9      4535.00    GTO  CCSFS SLC 40   
13            10  2014-09-07       Falcon 9      4428.00    GTO  CCSFS SLC 40   
14            11  2014-09-21       Falcon 9      2216.00    ISS  CCSFS SLC 40   
15            12  2015-01-10       Falcon 9      2395.00    ISS  CCSFS SLC 40   
16            13  2015-02-11       Falcon 9       570.00  ES-L1  CCSFS SLC 40   
17            14  2015-04-14       Falcon 9      1898.00    ISS  CCSFS SLC 40   
18            15  2015-04-27       Falcon 9      4707.00    GTO  CCSFS SLC 40   
19            16  2015-06-28       Falcon 9      2477.00    ISS  CCSFS SLC 40   
20            17  2015-12-22       Falcon 9      2034.00    LEO  CCSFS SLC 40   
21            18  2016-01-17       Falcon 9       553.00     PO   VAFB SLC 4E   
22            19  2016-03-04       Falcon 9      5271.00    GTO  CCSFS SLC 40   
23            20  2016-04-08       Falcon 9      3136.00    ISS  CCSFS SLC 40   
24            21  2016-05-06       Falcon 9      4696.00    GTO  CCSFS SLC 40   
25            22  2016-05-27       Falcon 9      3100.00    GTO  CCSFS SLC 40   
26            23  2016-07-18       Falcon 9      2257.00    ISS  CCSFS SLC 40   
27            24  2016-08-14       Falcon 9      4600.00    GTO  CCSFS SLC 40   
28            25  2016-09-01       Falcon 9      5500.00    GTO  CCSFS SLC 40   
29            26  2017-01-14       Falcon 9      9600.00     PO   VAFB SLC 4E   
30            27  2017-02-19       Falcon 9      2490.00    ISS    KSC LC 39A   
31            28  2017-03-16       Falcon 9      5600.00    GTO    KSC LC 39A   
32            29  2017-03-30       Falcon 9      5300.00    GTO    KSC LC 39A   
33            30  2017-05-01       Falcon 9          NaN    LEO    KSC LC 39A   
34            31  2017-05-15       Falcon 9      6070.00    GTO    KSC LC 39A   
35            32  2017-06-03       Falcon 9      2708.00    ISS    KSC LC 39A   
36            33  2017-06-23       Falcon 9      3669.00    GTO    KSC LC 39A   
37            34  2017-06-25       Falcon 9      9600.00     PO   VAFB SLC 4E   
38            35  2017-07-05       Falcon 9      6761.00    GTO    KSC LC 39A   
39            36  2017-08-14       Falcon 9      2910.00    ISS    KSC LC 39A   
40            37  2017-08-24       Falcon 9       475.00    SSO   VAFB SLC 4E   
41            38  2017-09-07       Falcon 9      4990.00    LEO    KSC LC 39A   
42            39  2017-10-09       Falcon 9      9600.00     PO   VAFB SLC 4E   
43            40  2017-10-11       Falcon 9      5200.00    GTO    KSC LC 39A   
44            41  2017-10-30       Falcon 9      3700.00    GTO    KSC LC 39A   
45            42  2017-12-15       Falcon 9      2205.00    ISS  CCSFS SLC 40   
46            43  2017-12-23       Falcon 9      9600.00     PO   VAFB SLC 4E   
47            44  2018-01-08       Falcon 9          NaN    LEO  CCSFS SLC 40   
48            45  2018-01-31       Falcon 9      4230.00    GTO  CCSFS SLC 40   
49            46  2018-03-06       Falcon 9      6092.00    GTO  CCSFS SLC 40   
50            47  2018-03-30       Falcon 9      9600.00     PO   VAFB SLC 4E   
51            48  2018-04-02       Falcon 9      2760.00    ISS  CCSFS SLC 40   
52            49  2018-04-18       

In [91]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9

/opt/conda/envs/Python-3.7-OpenCE/lib/python3.7/site-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


FlightNumber        Date BoosterVersion  PayloadMass  Orbit    LaunchSite  \
4              1  2010-06-04       Falcon 9          NaN    LEO  CCSFS SLC 40   
5              2  2012-05-22       Falcon 9       525.00    LEO  CCSFS SLC 40   
6              3  2013-03-01       Falcon 9       677.00    ISS  CCSFS SLC 40   
7              4  2013-09-29       Falcon 9       500.00     PO   VAFB SLC 4E   
8              5  2013-12-03       Falcon 9      3170.00    GTO  CCSFS SLC 40   
9              6  2014-01-06       Falcon 9      3325.00    GTO  CCSFS SLC 40   
10             7  2014-04-18       Falcon 9      2296.00    ISS  CCSFS SLC 40   
11             8  2014-07-14       Falcon 9      1316.00    LEO  CCSFS SLC 40   
12             9  2014-08-05       Falcon 9      4535.00    GTO  CCSFS SLC 40   
13            10  2014-09-07       Falcon 9      4428.00    GTO  CCSFS SLC 40   
14            11  2014-09-21       Falcon 9      2216.00    ISS  CCSFS SLC 40   
15            12  2015-01-10       Falcon 9      2395.00    ISS  CCSFS SLC 40   
16            13  2015-02-11       Falcon 9       570.00  ES-L1  CCSFS SLC 40   
17            14  2015-04-14       Falcon 9      1898.00    ISS  CCSFS SLC 40   
18            15  2015-04-27       Falcon 9      4707.00    GTO  CCSFS SLC 40   
19            16  2015-06-28       Falcon 9      2477.00    ISS  CCSFS SLC 40   
20            17  2015-12-22       Falcon 9      2034.00    LEO  CCSFS SLC 40   
21            18  2016-01-17       Falcon 9       553.00     PO   VAFB SLC 4E   
22            19  2016-03-04       Falcon 9      5271.00    GTO  CCSFS SLC 40   
23            20  2016-04-08       Falcon 9      3136.00    ISS  CCSFS SLC 40   
24            21  2016-05-06       Falcon 9      4696.00    GTO  CCSFS SLC 40   
25            22  2016-05-27       Falcon 9      3100.00    GTO  CCSFS SLC 40   
26            23  2016-07-18       Falcon 9      2257.00    ISS  CCSFS SLC 40   
27            24  2016-08-14       Falcon 9      4600.00    GTO  CCSFS SLC 40   
28            25  2016-09-01       Falcon 9      5500.00    GTO  CCSFS SLC 40   
29            26  2017-01-14       Falcon 9      9600.00     PO   VAFB SLC 4E   
30            27  2017-02-19       Falcon 9      2490.00    ISS    KSC LC 39A   
31            28  2017-03-16       Falcon 9      5600.00    GTO    KSC LC 39A   
32            29  2017-03-30       Falcon 9      5300.00    GTO    KSC LC 39A   
33            30  2017-05-01       Falcon 9          NaN    LEO    KSC LC 39A   
34            31  2017-05-15       Falcon 9      6070.00    GTO    KSC LC 39A   
35            32  2017-06-03       Falcon 9      2708.00    ISS    KSC LC 39A   
36            33  2017-06-23       Falcon 9      3669.00    GTO    KSC LC 39A   
37            34  2017-06-25       Falcon 9      9600.00     PO   VAFB SLC 4E   
38            35  2017-07-05       Falcon 9      6761.00    GTO    KSC LC 39A   
39            36  2017-08-14       Falcon 9      2910.00    ISS    KSC LC 39A   
40            37  2017-08-24       Falcon 9       475.00    SSO   VAFB SLC 4E   
41            38  2017-09-07       Falcon 9      4990.00    LEO    KSC LC 39A   
42            39  2017-10-09       Falcon 9      9600.00     PO   VAFB SLC 4E   
43            40  2017-10-11       Falcon 9      5200.00    GTO    KSC LC 39A   
44            41  2017-10-30       Falcon 9      3700.00    GTO    KSC LC 39A   
45            42  2017-12-15       Falcon 9      2205.00    ISS  CCSFS SLC 40   
46            43  2017-12-23       Falcon 9      9600.00     PO   VAFB SLC 4E   
47            44  2018-01-08       Falcon 9          NaN    LEO  CCSFS SLC 40   
48            45  2018-01-31       Falcon 9      4230.00    GTO  CCSFS SLC 40   
49            46  2018-03-06       Falcon 9      6092.00    GTO  CCSFS SLC 40   
50            47  2018-03-30       Falcon 9      9600.00     PO   VAFB SLC 4E   
51            48  2018-04-02       Falcon 9      2760.00    ISS  CCSFS SLC 40   
52            49  2018-04-18       

## Data Wrangling


We can see below that some of the rows are missing values in our dataset.


In [94]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        5
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Before we can continue we must deal with these missing values. The <code>LandingPad</code> column will retain None values to represent when landing pads were not used.


### Task 3: Dealing with Missing Values


Calculate below the mean for the <code>PayloadMass</code> using the <code>.mean()</code>. Then use the mean and the <code>.replace()</code> function to replace `np.nan` values in the data with the mean you calculated.


In [97]:
# Calculate the mean value of PayloadMass column
data_falcon9['PayloadMass'].fillna(value=data_falcon9['PayloadMass'].mean(), inplace=True)
# Replace the np.nan values with its mean value

False

You should see the number of missing values of the <code>PayLoadMass</code> change to zero.


Now we should have no missing values in our dataset except for in <code>LandingPad</code>.


We can now export it to a <b>CSV</b> for the next section,but to make the answers consistent, in the next lab we will provide data in a pre-selected date range.


<code>data_falcon9.to_csv('dataset_part\_1.csv', index=False)</code>


In [98]:
data_falcon9.to_csv('dataset_part_01.csv', index=False)

## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork26802033-2021-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.


## Change Log


| Date (YYYY-MM-DD) | Version | Changed By | Change Description                  |
| ----------------- | ------- | ---------- | ----------------------------------- |
| 2020-09-20        | 1.1     | Joseph     | get result each time you run        |
| 2020-09-20        | 1.1     | Azim       | Created Part 1 Lab using SpaceX API |
| 2020-09-20        | 1.0     | Joseph     | Modified Multiple Areas             |


Copyright © 2021 IBM Corporation. All rights reserved.
